### PR TITLE
Fix shared object versioning

### PR DESCRIPTION
**Pull request by kmilos**
This now decorrelates shared object (exiv2lib) version from project/package version, [as a potential solution discussed](https://github.com/Exiv2/exiv2/issues/2020#issuecomment-990851749).

To continue the pre-existing sequence, the SOVERSION is now set at 28. A (small) side effect without consequence (I think, famous last words), is that the library version will go from e.g. 0.27.6 to 28.0.0 now. So today, on *nix one has a symlink like this

[libexiv2.so.27 -> libexiv2.so.0.27.6](https://packages.debian.org/sid/amd64/libexiv2-27/filelist)

which will change to

libexiv2.so.28 -> libexiv2.so.28.0.0

Currently main produces

libexiv2.so.0 -> libexiv2.so.1.0.0

which is just wrong (SOVERSION decreases from previous version).

Btw, this is e.g. how libtiff and libavif do it among many others...

### CODE REVIEW

**Review by kevinbackhouse — DISMISSED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,16 @@ project(exiv2          # use TWEAK to categorize the build
...
 )
 
+# Shared Object versioning (simplified Libtool scheme):
+#   If the source code has changed, but there were no interface changes, increment REVISION.
+#   If there was a compatible interface change, increment MINOR and set REVISION back to 0.
+#   If there was an incompatible interface change, increment MAJOR, and set MINOR and REVISION back to 0.
+set(EXIV2LIB_MAJOR 28)
+set(EXIV2LIB_MINOR 0)
+set(EXIV2LIB_REVISION 0)
+set(EXIV2LIB_VERSION "${EXIV2LIB_MAJOR}.${EXIV2LIB_MINOR}.${EXIV2LIB_REVISION}")
```
Comment (kmilos) [inline CMakeLists.txt:22]: Another option could be `set(EXIV2LIB_VERSION "${PROJECT_VERSION_MAJOR}.${EXIV2LIB_MAJOR}.${EXIV2LIB_MINOR}.${EXIV2LIB_REVISION}")`?

**Comment by piponazo** [comment]
Ey hi, I hope I am not arriving too late to the discussion. 

I am wondering why to create some new variables instead of keep using the CMake ones defined by the **project()** command. Basically by changing the version in line 5 to 28.0.0 we would achieve the same effect. Or am I missing something?

**Comment by kmilos** [comment]
@piponazo There is nothing that says that a library interface number (or compatibility version in macOS speak) has to be the same or in any way derived from the project/package release number, and for many libraries it is the case.

It is of course simpler if they are. However, our library interface is now at version 27, and future ones must increase. Unless we continue w/ releases like 0.28.x, 0.29.x, 0.30.x, ad infinitum, how do you propose we solve this conundrum exactly?

**Comment by kmilos** [comment]
> Basically by changing the version in line 5 to 28.0.0 we would achieve the same effect. Or am I missing something?

I don't think we want to go from releasing a 0.27.6 package to a 28.0.0 package 😉

**Comment by piponazo** [comment]
> @piponazo There is nothing that says that a library interface number ("soversion" or compatibility version in macOS speak) has to be the same or in any way derived from the project/package release number, and for many libraries this is the case. The interfaces change, more often slower, than package releases.

Ok, I see your point. What I do not understand yet is why we want to complicate our lives and maintain 2 different set of version variables (although it seems that it is related with the hope of releasing a 1.0 version at some point in the future).

By the way, in my current job we recently started to use the GitVersion tool in CI which takes care of generating nice SemanticVersion numbers depending on the latest tag of the repository and the branches in which you work:
https://gitversion.net/docs/

If your project sticks to some well known branching strategy, it works pretty well.

**Comment by kmilos** [comment]
> What I do not understand yet is why we want to complicate our lives and maintain 2 different set of version variables (although it seems that it is related with the hope of releasing a 1.0 version at some point in the future).

Our lives are already complicated by the current state of things and choices made previously. Some people using main are already bitten by the fact that main has backtracked from SOVERSION of 27 to 0 (which also might have existed at some point in the distant past)...

Any concrete proposals how to move us forward?

Although I'm ok with completely decoupled project and library versions, here's a few more suggestions in pseudocode that would let you keep only the one (inspired by Robin's ideas in the linked issue):

if PROJECT_VERSION_MAJOR = 0
  EXIV2LIB_SOVERSION = PROJECT_VERSION_MINOR
else
  EXIV2LIB_SOVERSION = PROJECT_VERSION_MAJOR + 27 (assuming we don't do a 0.28, or use whatever offset we stop at before 1.0)
                                       or = 100 * PROJECT_VERSION_MAJOR (e.g. jump from 27/28 to 100 next)

or 

EXIV2LIB_SOVERSION = 100 * PROJECT_VERSION_MAJOR + PROJECT_VERSION_MINOR


These possibly let us leave

EXIV2LIB_VERSION = PROJECT_VERSION

**Comment by neheb** [comment]
meson.build should be changed as well

**Comment by kmilos** [comment]
> meson.build should be changed as well

Sure, once there is a clear agreement on the way forward. @nehaljwani

**Review by nehaljwani — DISMISSED** [review]

**Comment by kmilos** [comment]
Thanks for taking a look @nehaljwani - does this mean there is no intention of ever doing a 0.28.x?

**Comment by nehaljwani** [comment]
>EXIV2LIB_SOVERSION = 100 * PROJECT_VERSION_MAJOR + PROJECT_VERSION_MINOR
>
>These possibly let us leave
>
>EXIV2LIB_VERSION = PROJECT_VERSION

IMHO, this is the best path forward. @kmilos could you please update the pull request?

**Comment by nehaljwani** [comment]
Sorry, I deleted my last comment. I had the incorrect block in quotes.

>if PROJECT_VERSION_MAJOR = 0
>  EXIV2LIB_SOVERSION = PROJECT_VERSION_MINOR
>else
>  EXIV2LIB_SOVERSION = 100 * PROJECT_VERSION_MAJOR

I like this approach. Sure, it will be a big number with the next major release, but I don't think it matters, as only linkers are gonna care about it. We can also drop the if/else and just keep the second branch once we cut the major release.

**Comment by nehaljwani** [comment]
> Thanks for taking a look @nehaljwani - does this mean there is no intention of ever doing a 0.28.x?

I'm very less inclined to, but at least with the block quoted above, we'll keep the option open.

**Comment by kmilos** [comment]
Ok, I updated the PR now, restarting the SOVERSION from 30 for 1.0.0 ;) (this still give the opportunity to have 0.28 and 0.29 series)

@neheb `soversion` option also needs to be added to Meson [here](https://github.com/Exiv2/exiv2/blob/3403dabec247a41206138645b5045fd2cfddf2a3/meson.build#L303-L314), but I don't know the arithmetic syntax, can you please contribute the equivalent snippet once this is accepted?

**Review by nehaljwani — APPROVED** [review]
I'm fine with the proposed change.

**Comment by neheb** [comment]
A similar change is needed for the meson build.

**Comment by kmilos** [comment]
I  did it for meson?

**Comment by neheb** [comment]
Missed that commit. Great.

**Pull request merged**